File: machine/ts/render/vdp/framebuffer.ts
import type { TextureHandle } from '../backend/backend';
import type { VDP } from '../../machine/devices/vdp/vdp';
import { FRAMEBUFFER_RENDER_TEXTURE_KEY, FRAMEBUFFER_TEXTURE_KEY } from '../../rompack/format';
import { RGBA8_LINEAR_TEXTURE_PARAMS } from '../backend/texture_params';
import type { TextureManager } from '../texture_manager';
import type { GameView } from '../gameview';

const EMPTY_TEXTURE_SEED = new Uint8Array(4);

export class VdpFrameBufferTextures {
	private renderFrameBufferTexture: TextureHandle = null as TextureHandle;
	private displayFrameBufferTexture: TextureHandle = null as TextureHandle;
	private frameBufferTextureWidth = 0;
	private frameBufferTextureHeight = 0;

	public constructor(
		private readonly textureManager: TextureManager,
		private readonly view: GameView,
	) {
	}

	public initialize(vdp: VDP): void {
		this.frameBufferTextureWidth = vdp.frameBufferWidth;
		this.frameBufferTextureHeight = vdp.frameBufferHeight;
		this.renderFrameBufferTexture = this.textureManager.createTextureFromPixelsSync(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			EMPTY_TEXTURE_SEED,
			1,
			1,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
		this.renderFrameBufferTexture = this.textureManager.resizeTextureForKey(
			FRAMEBUFFER_RENDER_TEXTURE_KEY,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = this.renderFrameBufferTexture;
		this.displayFrameBufferTexture = this.textureManager.createTextureFromPixelsSync(
			FRAMEBUFFER_TEXTURE_KEY,
			EMPTY_TEXTURE_SEED,
			1,
			1,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
		this.displayFrameBufferTexture = this.textureManager.resizeTextureForKey(
			FRAMEBUFFER_TEXTURE_KEY,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
		this.view.textures[FRAMEBUFFER_TEXTURE_KEY] = this.displayFrameBufferTexture;
		this.view.backend.updateTextureRegion(
			this.textureManager.getTextureByUri(FRAMEBUFFER_RENDER_TEXTURE_KEY, RGBA8_LINEAR_TEXTURE_PARAMS),
			vdp.frameBufferRenderReadback,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			0,
			0,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
		this.view.backend.updateTextureRegion(
			this.textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, RGBA8_LINEAR_TEXTURE_PARAMS),
			vdp.frameBufferDisplayReadback,
			vdp.frameBufferWidth,
			vdp.frameBufferHeight,
			0,
			0,
			RGBA8_LINEAR_TEXTURE_PARAMS
		);
	}

	public width(): number {
		return this.frameBufferTextureWidth;
	}

	public height(): number {
		return this.frameBufferTextureHeight;
	}

	public displayTexture(): TextureHandle {
		return this.displayFrameBufferTexture;
	}

	public renderTexture(): TextureHandle {
		return this.renderFrameBufferTexture;
	}
}

File: machine/cpp/render/vdp/framebuffer.cpp
#include "render/vdp/framebuffer.h"

#include "machine/devices/vdp/vdp.h"
#include "render/gameview.h"
#include "render/texture_manager.h"
#include "rompack/format.h"
#include <array>

namespace bmsx {
namespace {

const std::array<u8, 4> EMPTY_TEXTURE_SEED{{0, 0, 0, 0}};
} // namespace

VdpFrameBufferTextures::VdpFrameBufferTextures(TextureManager& textureManager, GameView& view)
	: m_textureManager(textureManager)
	, m_view(view) {
}


void VdpFrameBufferTextures::initialize(VDP& vdp) {
	m_frameBufferTextureWidth = vdp.frameBufferWidth();
	m_frameBufferTextureHeight = vdp.frameBufferHeight();
	m_renderFrameBufferTexture = m_textureManager.createTextureFromPixelsSync(
		FRAMEBUFFER_RENDER_TEXTURE_KEY,
		EMPTY_TEXTURE_SEED.data(),
		1,
		1,
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
	m_renderFrameBufferTexture = m_textureManager.resizeTextureForKey(
		FRAMEBUFFER_RENDER_TEXTURE_KEY,
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
	m_view.textures[FRAMEBUFFER_RENDER_TEXTURE_KEY] = m_renderFrameBufferTexture;
	m_displayFrameBufferTexture = m_textureManager.createTextureFromPixelsSync(
		FRAMEBUFFER_TEXTURE_KEY,
		EMPTY_TEXTURE_SEED.data(),
		1,
		1,
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
	m_displayFrameBufferTexture = m_textureManager.resizeTextureForKey(
		FRAMEBUFFER_TEXTURE_KEY,
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
	m_view.textures[FRAMEBUFFER_TEXTURE_KEY] = m_displayFrameBufferTexture;
	m_view.backend()->updateTextureRegion(
		m_textureManager.getTextureByUri(FRAMEBUFFER_RENDER_TEXTURE_KEY, RGBA8_LINEAR_TEXTURE_PARAMS),
		vdp.frameBufferRenderReadback().data(),
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		0,
		0,
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
	m_view.backend()->updateTextureRegion(
		m_textureManager.getTextureByUri(FRAMEBUFFER_TEXTURE_KEY, RGBA8_LINEAR_TEXTURE_PARAMS),
		vdp.frameBufferDisplayReadback().data(),
		static_cast<i32>(m_frameBufferTextureWidth),
		static_cast<i32>(m_frameBufferTextureHeight),
		0,
		0,
		RGBA8_LINEAR_TEXTURE_PARAMS
	);
}

} // namespace bmsx
